zingo-sync/src/sync.rs

```diff
@@ -27,62 +31,107 @@ pub async fn sync<P, W>(
     wallet: &mut W,
 ) -> Result<(), ()>
 where
-    P: Parameters + Send + 'static,
-    W: SyncWallet,
+    P: Parameters + Sync + Send + 'static,
+    W: SyncWallet + SyncBlocks + SyncNullifiers + SyncShardTrees,
 {
     tracing::info!("Syncing wallet...");
 
-    // TODO: add trait methods to read/write wallet data to/from sync engine
-    // this is where sync state would be read from wallet data
-    let sync_state = SyncState::new(); // placeholder
+    let mut handles = Vec::new();
 
     // create channel for sending fetch requests and launch fetcher task
-    let (fetch_request_sender, fetch_request_receiver) = unbounded_channel();
+    let (fetch_request_sender, fetch_request_receiver) = mpsc::unbounded_channel();
     let fetcher_handle = tokio::spawn(fetch(fetch_request_receiver, client));
+    handles.push(fetcher_handle);
+
+    update_scan_ranges(
+        fetch_request_sender.clone(),
+        parameters,
+        wallet.get_birthday().unwrap(),
+        wallet.get_sync_state_mut().unwrap(),
+    )
+    .await
+    .unwrap();
+
+    // create channel for receiving scan results and launch scanner
+    let (scan_results_sender, mut scan_results_receiver) = mpsc::unbounded_channel();
+    let ufvks = wallet.get_unified_full_viewing_keys().unwrap();
+    let mut scanner = Scanner::new(
+        scan_results_sender,
+        fetch_request_sender,
+        parameters.clone(),
+        ufvks,
+    );
+    scanner.spawn_workers();
+
+    let mut interval = tokio::time::interval(Duration::from_millis(30));
+    loop {
+        interval.tick().await;
+
+        // if a scan worker is idle, send it a new scan task
+        if scanner.is_worker_idle() {
+            if let Some(scan_range) = prepare_next_scan_range(wallet.get_sync_state_mut().unwrap())
+            {
+                let previous_wallet_block = wallet
+                    .get_wallet_block(scan_range.block_range().start - 1)
+                    .ok();
+
+                scanner
+                    .add_scan_task(ScanTask::from_parts(scan_range, previous_wallet_block))
+                    .unwrap();
+            } else {
+                // when no more ranges are available to scan, break out of the loop
+                break;
+            }
+        }
+
+        match scan_results_receiver.try_recv() {
+            Ok((scan_range, scan_results)) => {
+                update_wallet_data(wallet, scan_results).unwrap();
+                // TODO: link nullifiers and scan linked transactions
+                remove_irrelevant_data(wallet, &scan_range).unwrap();
+                mark_scanned(scan_range, wallet.get_sync_state_mut().unwrap()).unwrap();
+            }
+            Err(TryRecvError::Empty) => (),
+            Err(TryRecvError::Disconnected) => break,
+        }
+    }
```

Comment:
at this point IIUC, either the connection broke or all ranges are scanned.
are we now making assumptions about how many tuples are left in the receiver? couldnt it still be more than one? or shouldnt it be 0?